Disconnect / reconnect edge source and target by clicking on edges. (Full)

MichaelMauderer opened the pull request:
### Pull Request Description
Implement full edge splitting.

### Important Notes
Note comments and discussions from the previous partial PR: https://github.com/enso-org/ide/pull/574

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [x] All code has been tested where possible.
- [x] All code has been profiled where possible.

wdanilo commented:
Hi, amazing progress! It works smoothly and a lot of issues are gone. Still, there are few:

1. Clicking on edge disconnects the wrong part:
    ![2020-07-08 10 41 05](https://user-images.githubusercontent.com/1623053/86897519-cad35200-c107-11ea-843a-18d2937a463f.gif)
2. Clicking on output port and then input port creates a new edge while it should not.
    ![2020-07-08 10 43 24](https://user-images.githubusercontent.com/1623053/86897772-269ddb00-c108-11ea-94d1-b000629e37ec.gif)
3. You can ... connect edge to edge now, and even create edge loop!
    ![2020-07-08 10 44 30](https://user-images.githubusercontent.com/1623053/86897839-3c130500-c108-11ea-9d47-3e2292ce490f.gif)
4. There is edge precedence error, which is strange. When dragging an edge, all of its parts should e drawn in front layer (at least this is behavior before this PR). Now, its not the case anymore. It works for output-only attached edges, but not for input-only attached ones:
    ![2020-07-08 10 46 29](https://user-images.githubusercontent.com/1623053/86898044-7aa8bf80-c108-11ea-9263-885dff6832a0.gif)
    ![2020-07-08 10 48 33](https://user-images.githubusercontent.com/1623053/86898195-ae83e500-c108-11ea-94c5-2049dd796ed5.gif)
5. There are some strange "dots" appearing around arrow when hovering the edge:
    ![2020-07-08 10 52 04](https://user-images.githubusercontent.com/1623053/86898555-26eaa600-c109-11ea-953e-875f7de48b92.gif)
6. The arrow top has strange artifacts when hovering:
    ![2020-07-08 10 53 22](https://user-images.githubusercontent.com/1623053/86898680-50a3cd00-c109-11ea-9969-bb5085ab2019.gif)
7. The division between top and bottom edge part is not in the middle of the nodes:
    ![2020-07-08 10 55 00](https://user-images.githubusercontent.com/1623053/86898891-8e085a80-c109-11ea-9f04-5b279800f38a.gif)
8. Plus, create even wider hover area for edges please, so this situation would also hover the edge:
    ![image](https://user-images.githubusercontent.com/1623053/86898411-f440ad80-c108-11ea-8c80-10f9b1f965cb.png)

wdanilo commented:
I was checking the status of today implementation and here are I think new cases I found:

9. Desynchronised position on hover when zooed in
    ![2020-07-09 02 18 39](https://user-images.githubusercontent.com/1623053/86983185-77a2e300-c18b-11ea-8699-0f6214d539e3.gif)

10. This glitch:
    ![2020-07-09 02 19 48](https://user-images.githubusercontent.com/1623053/86983190-7c679700-c18b-11ea-8d1a-5139bf694c4a.gif)

11. This actually should just not be possible (and I think we overlooked it) - when dragging edge with only output port connected, the mouse should not get snapped to inputs because it is impossible to create input <-> input connection and it is deeply confusing.
    ![2020-07-09 02 20 55](https://user-images.githubusercontent.com/1623053/86983191-7d002d80-c18b-11ea-9a34-87ff880ae19f.gif)

12. When starting a connection when zoomed, the connection is drawn not to the mouse position if mouse was not yet moved.
    ![2020-07-09 02 22 04](https://user-images.githubusercontent.com/1623053/86983192-7d98c400-c18b-11ea-9289-dc714a05744d.gif)

13. After starting a connection it is impossible to click in the background to drop it, probably because of the glitches that are seen there. I suspect this might be issue introduced by my connection implementation and handling corner cases of very short connections - so Im documenting it here, but if this is my issue (please check on aster if it behaves the same way for short connections), I'll fix it as another ticket.
    ![2020-07-09 02 22 30](https://user-images.githubusercontent.com/1623053/86983193-7e315a80-c18b-11ea-9089-b8375305695a.gif)

14. This glitch
    ![2020-07-09 02 23 57](https://user-images.githubusercontent.com/1623053/86983194-7ec9f100-c18b-11ea-828a-f9cfb1dcd37d.gif)

wdanilo commented:
Plus, I was browsing the code fast (not real review yet), and I have a few comments - please make sure that:
1. You do not use `.display_object()` explicitly.
2. Spacing is correct (between sections, function args, etc)
3. I'd be thankful if you make sure that all docs of the abstractions you introduce would contain clear descriptions regarding **what they are used for**. So for example, `SemanticSplit` - I see what it does, but I was unable to really answer the question of why this abstraction has to exist and what it really is used for. Such info in the documentation would drastically help understand what's going.

MichaelMauderer commented:
@wdanilo 
* 1 fixed
* 2 fixed, but we should also set up an issue to do the correct edge handling for the multi-output port shapes.
* 3 fixed
* 4 fixed
* 5 fixed
* 6 fixed 
* 7 fixed 
* 8 fixed 
* 9 This is also on `main` and I will add this to a new issue about zooming breaking interactions (#635). 
* 10 fixed
* 11 This is partially fixed but comes at the trade-off that either we have no snapping when not doing edge dragging or the behavior you describe. A full fix should go into a bugfix issue because it will need some additional work on handling the cursor style from the port manager (#659). 
* 12 fixed
* 13 The dropping of edges seems by clicking on the background seems to work for me. I will open a new issue for the shape glitch.  (#658)
* 14 fixed
* 15 fixed

wdanilo commented:
Hi! I'm happy it's getting polished, but unfortunately, a set of new bugs appeared (these are regressions as were not happening earlier as far as I know):

16. It is impossible now to create multiple output edges - clicking on output port when an output edge exists, does nothing.
17. There is a new glitch (was not present before):
    ![2020-07-16 09 29 01](https://user-images.githubusercontent.com/1623053/87640347-dbf31300-c746-11ea-875d-f42de8274226.gif)
18. The edge highlighting is inverted - it just should work other way around. When mouse is close to the input port and you click on the edge, the part of the edge close to the port should stay active, not the second part. I realized that after seeing issue 19 (below).
19. When input is connected, clicking on it does nothing. Instead, when you click on input port, it should just disconnect existing edge (it should be consistent with the edge highlight) - when you move over the edge, and then you hover the port, the input port should behave just like "the last part of the edge".
20. This happens now (this makes me especially worried as it looks like you are keeping a mutable-like state on the edges, which you should not):
    ![2020-07-16 09 57 10](https://user-images.githubusercontent.com/1623053/87643211-d7305e00-c74a-11ea-958b-a5fd7e5ebacb.gif)
21. I cannot connect edges anymore in this configuration (moreover, a lot of "jumping" is seen here, as in the issue 20):
    ![2020-07-16 09 57 21](https://user-images.githubusercontent.com/1623053/87643272-edd6b500-c74a-11ea-81f1-07af0116c777.gif)
22. This glitch (happens with 2 nodes connected together too):
    ![2020-07-16 10 02 58](https://user-images.githubusercontent.com/1623053/87643842-a866b780-c74b-11ea-8e7c-c33506050caf.gif)
22. This glitch (I bet it was reported already, but cannot find the number for it):
    ![2020-07-16 09 37 54](https://user-images.githubusercontent.com/1623053/87644758-d993b780-c74c-11ea-89c7-ef38561e5119.gif)

**Moreover, few issues reported earlier are not yet fixed. Please be sure to fix and TEST all reported issues before making a PR - this would save us a lot of time. In particular:**

4. Still seeing the glitch:
    ![2020-07-16 09 41 47](https://user-images.githubusercontent.com/1623053/87641643-afd89180-c748-11ea-9815-611939b9ae8f.gif)
7. The division point is still not in the middle of the distance between nodes, see the attached image - the mouse is placed at the division point, and the rectangles have the same height:
    <img width="160" alt="rects" src="https://user-images.githubusercontent.com/1623053/87642232-866c3580-c749-11ea-9382-3425d4ebe07b.png">
9. This still happens - but it's ok, as you described, let's make another issue from that.
11. This issue still exists - ok, we can make it a separate issue, but the solution for that will be simple - I believe a simple FRP gate would do the trick. But, let's keep it separate.
15. It's not fixed:
    ![2020-07-16 10 07 47](https://user-images.githubusercontent.com/1623053/87644327-48bcdc00-c74c-11ea-8028-dfbdc7372b73.gif)

MichaelMauderer commented:
@wdanilo 

**5** Fixed. According to my measurements, this is halfway between the edges of the nodes now. (Which means one line is longer, as it overlaps the node)
**16** Fixed.
**4 + 8 + 17 + 20**: This is all related to my initial fix for (4) and shows that there is more work to be done on changing the shape of the edge to have the correct ordering behavior in all those cases. Since this is not really related to edge disconnecting I have opened issue #672 to deal with those. I've added what I found out about how to resolve the issues there, too.
**18 + 19** This should be a new issue, as it is new behavior unrelated to the split/highlight.disconnect of this PR (and I want to avoid growing this PR even more).
**21 + 8r:** Fixed. (The original 8 was a different underlying issue) There is a trade-off here: the arrow must only appear if the mainline is as high as the arrow. Otherwise we end up in a bad shape configuration where there is a discontinuity in the line.
**22** Fixed by limiting the snapping to the visible circle quadrant.
**23** Fixed. Was caused by hovering information being used when detached. Fixed by not allowing the hover split when edge is detached.

MichaelMauderer commented:
One more thing that needs fixing before re-review though

![Peek 2020-07-17 12-48](https://user-images.githubusercontent.com/1428930/87778683-de7b6880-c82b-11ea-83d9-777b295e998e.gif)

MichaelMauderer commented:
@wdanilo This is ready for another look.

wdanilo commented:
Michael, I'm happy to see it ats this stage, good work! Edges are fun :) 

Replying to above message:

- **4 + 17 + 20** Ok, lets do it as separate issue. I have linked it to https://github.com/enso-org/ide/issues/570.
- **18** I do not agree this is a separate issue. The current highlighting is wrong - it is inverted. Just invert it - highlight the other part of the edge.
- **19** I agree this could be a separate issue – please create it and link it to https://github.com/enso-org/ide/issues/570


Moreover, I found new things:

24. The height of the edge jumps when hovering input port - it was not behaving like that:
    ![2020-07-20 19 53 31](https://user-images.githubusercontent.com/1623053/87972181-c4a98200-cac6-11ea-878d-938f53c90032.gif)

25. This glitch is similar to 15 (blinking of the edge) - to reproduce it - place the left node just SLIGHTLY below the right node:
    ![2020-07-20 20 32 05](https://user-images.githubusercontent.com/1623053/87973104-506fde00-cac8-11ea-8268-fcec8efe822f.gif)

16. This is pretty bad - if nodes are on the same Y-axis, its impossible to disconnect one of them. HOWEVER, let's make it part of https://github.com/enso-org/ide/issues/570 as well.



Moreover, I found a new thing that is not implemented but should be part of this task. However, as I discovered it really late, let's make it part of https://github.com/enso-org/ide/issues/570. The thing is that we should support 2 modes of disconnecting - clicking, and press-nad-drag mode, just like on input / output ports. Currently, edges support only first mode.


**So to sum this up, please fix 18, and 24, 25, and then let's merge it :)**

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/enso-org/ide?pullRequest=640) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/enso-org/ide?pullRequest=640) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: MichaelMauderer<br/>:x: wdanilo<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/enso-org/ide?pullRequest=640) it.</sub>

wdanilo commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -1830,29 +1966,38 @@ fn new_graph_editor(app:&Application) -> GraphEditor {
-        edges.detached_target.for_each(|id| {
+    refresh_target      <- any(&edge_refresh_cursor_pos_on_hover,&edge_refresh_cursor_pos_no_hover);
+    refresh_source      <- edge_refresh_cursor_pos_no_hover.map(|pos| *pos);
+    snap_source_to_node <- edge_refresh_on_node_hover.map(|(_,target)| target.clone());
```
you can use `._1()` instead

wdanilo commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -1830,29 +1966,38 @@ fn new_graph_editor(app:&Application) -> GraphEditor {
-    eval edge_refresh_cursor_pos ((position) {
-        edges.detached_target.for_each(|id| {
+    refresh_target      <- any(&edge_refresh_cursor_pos_on_hover,&edge_refresh_cursor_pos_no_hover);
+    refresh_source      <- edge_refresh_cursor_pos_no_hover.map(|pos| *pos);
```
what does this line do? If im correct, this does not do anything.

MichaelMauderer replied:
This is mostly for naming ourpooses to make clear how this is used next.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -1868,7 +2013,7 @@ fn new_graph_editor(app:&Application) -> GraphEditor {
     });
 
-    eval edge_refresh_on_node_hover ([nodes,edges,model]((_,target)) {
+    eval snap_source_to_node ([nodes,edges,model]((target)) {
```
unnecesary parenthesis - this should be catched by linter btw

MichaelMauderer replied:
Linter does not complain but it is fixed.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+
+
+
+fn up() -> Vector2<f32> {
+    Vector2(1.0,0.0)
+}
+
+fn point_rotation(point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+    nalgebra::Rotation2::rotation_between(&point,&up())
+}
+
```
I know it was my change, but I did not clean my code. This should be in a section imo

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+/// Add an invisible hover area to the provided shape. The base shape should already be colored
+/// otherwise coloring it later will also color the hover area.
+fn hover_area(base_shape:AnyShape, size:Var<Pixels>) -> AnyShape {
+    let hover_area = base_shape.grow(size).fill(color::Rgba::new(0.0,0.0,0.0,0.000_001));
```
this sohuld use the color constant defined somewhere above

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+    ensogl::define_shape_system! {
+            (
+            ) {
+                let radius : Var<Pixels> = "min(input_size.y,input_size.x)".into();
```
wouldn't these be always the same ? Is there any place where we keep the input_size not rectangular? If no, let's not waste gpu computing power ;)

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
-                let n_shape  = Rect((&shadow_size*2.0 + 2.px() * dim.x(),&n_radius*2.0)).corners_radius(n_radius);
-                let n_shape  = n_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
+                let node_radius = &shadow_size + 1.px() * dim.y();
+                let node_shape  = Rect(
+                    (&shadow_size*2.0 + 2.px() * dim.x(),&node_radius*2.0)).corners_radius(node_radius);
```
just make vars `node_width` and `node_height` :)

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+                let node_radius = &shadow_size + 1.px() * dim.y();
+                let node_shape  = Rect(
+                    (&shadow_size*2.0 + 2.px() * dim.x(),&node_radius*2.0)).corners_radius(node_radius);
+                let node_shape  = node_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
```
again, this is refactored somewhere to constants. Lets replace all such colors by the constant.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+                point_rotation(point)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                let radius = self.shape.radius.get();
+                let center = Vector2::zero();
+                let point_to_center = point.xy() - center;
+
+                let closest_point = center + point_to_center / point_to_center.magnitude() * radius;
+                let vector_angle  = -nalgebra::Rotation2::rotation_between(&Vector2::new(0.0, 1.0),&closest_point).angle();
+                let start_angle   =  self.shape.start_angle.get();
+                let end_angle     =  start_angle + self.shape.angle.get();
+                let upper_bound   = start_angle.max(end_angle);
+                let lower_bound   = start_angle.min(end_angle);
+
+                let correct_quadrant = lower_bound < vector_angle && upper_bound > vector_angle;
+                if correct_quadrant {
+                     Some(Vector2::new(closest_point.x, closest_point.y))
+                } else {
+                    None
+                }
```
As we've been discussing - this is too complex and this is just a hotfix. There should be TODO note about it with link to an issue about it before this PR is merged. (the TODO should be in every such place)

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+
+                let closest_point = center + point_to_center / point_to_center.magnitude() * radius;
+                let vector_angle  = -nalgebra::Rotation2::rotation_between(&Vector2::new(0.0, 1.0),&closest_point).angle();
+                let start_angle   =  self.shape.start_angle.get();
+                let end_angle     =  start_angle + self.shape.angle.get();
```
2 spaces after `=`

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                let radius = self.shape.radius.get();
```
why this does not need to have allow(depraecated)?

MichaelMauderer replied:
Because we always were able to get the values from shapes. Just not sprites. Should this be different?

wdanilo replied:
Interesting. I don't know. I will think about it. Thanks for pointing it out!

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                #[allow(deprecated)]
```
all such places needs to be propelry documented. No one will remember to fix that in 1 month from now without proper documentation and link to an issue.

wdanilo commented on src/rust/ensogl/src/display/symbol/gpu/geometry/compound/sprite.rs:
```diff
@@ -119,6 +119,18 @@ impl Size {
+
+#[allow(missing_docs)]
+impl Size {
+    #[deprecated(note = "This will be removed without replacement soon.")]
+    pub fn get(&self) -> Vector2<f32>{
+        if self.hidden.get() { self.value.get() }
+        else                 { self.attr.get()  }
+    }
+}
```
Let's add more context here. Someone reading such things should understand WHY it is deprecated. We want to keep here `Var`s instead of normal values, so `get` will not result in value soon.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -960,10 +960,69 @@ impl GraphEditorModelWithNetwork {
+    fn new_edge_from_input
+     (&self,edge_click:&frp::Source<EdgeId>,edge_over:&frp::Source<EdgeId>,
+      edge_out:&frp::Source<EdgeId>) -> EdgeId {
+        let edge_id = self.create_edge(edge_click,edge_over,edge_out);
+        let first_detached = self.edges.detached_source.is_empty();
```
alignment -- in other places as well, please make a fast self-review on github and check for all such places.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -960,10 +960,69 @@ impl GraphEditorModelWithNetwork {
+    }
+
+    fn new_edge_from_input
+     (&self,edge_click:&frp::Source<EdgeId>,edge_over:&frp::Source<EdgeId>,
+      edge_out:&frp::Source<EdgeId>) -> EdgeId {
```
in such case please use param by line

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+                let lower_bound   = start_angle.min(end_angle);
+
+                let correct_quadrant = lower_bound < vector_angle && upper_bound > vector_angle;
+                if correct_quadrant {
+                     Some(Vector2::new(closest_point.x, closest_point.y))
+                } else {
+                    None
+                }
```
in such cases use always `as_some` from boolinator exported in proelude

MichaelMauderer replied:
What is the advantage of that over the explicit `Some(...)`?
Besides I am always getting a not-implemented error for `as_some`. Need to investigate why.

wdanilo replied:
Because the code is much nicer!

```rust
if correct_quadrant {
                     Some(Vector2::new(closest_point.x, closest_point.y))
                } else {
                    None
                }
```
vs
```rust
correct_quadrant.as_some(Vector2::new(closest_point.x,closest_point.y))
```

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
-                let shape      = shape.fill(color::Rgba::from($($color)*));
+                let focus_split_angle  = focus_split_angle.into();
+                let focus_split_center = focus_split_center.px();
+                let shape = Triangle(width-1.0.px(),height-1.0.px());
```
what are these `-1.px()` ? Btw you can write `1.px()` instead of `1.0.px()`. They should be named variable telling what they are for.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -251,23 +672,45 @@ pub mod back {
+                $(let $field = component::ShapeView::new(
+                    Logger::sub(&logger,stringify!($field)),scene);)*
+                    $(display_object.add_child(&$field);)*
+                let mut shape_view_events:Vec<ShapeViewEvents> = Vec::default();
```
a very very minor thing. Just ocnidering our alignment rules, it's always better to write such things as `= Vec::<ShapeViewEvents>::default()` - then left side of `=` is shorter. Doesnt matter here but I thought I will mention it.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+    fn is_input_above_output(self) -> bool {
+        !self.is_output_above_input()
+    }
+
```
rm nl

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+/// Note that "at the split" also includes the shapes adjacent to the actual split because they
+/// need to be treated as if they were at the split location to avoid glitches at the shape
+/// boundaries.
+/// // FIXE: please, tell more about these glitches - I do not understand why they happen after this description
```
please fix this fixme :) OH I MADE TYPO HERE! Sorry :)

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+        Some(SemanticSplit {ordered_part_ids,split_index})
+    }
+
+    /// Return a ordered vector that contains the ids of the shapes in the order they appear in the
```
a -> an

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+    }
+
+    /// Return a ordered vector that contains the ids of the shapes in the order they appear in the
+    /// edge. Shapes that are to be handled as in the same place, are binned into a sub-vector.
```
Are you sure you wanted to use the "binned" verb here? My diet tells it means that something is put in a bin, but this could also be my weak english

MichaelMauderer replied:
They are put in a bin :) This is derived from [data binning/bucketing](https://en.wikipedia.org/wiki/Data_binning)

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+            LayoutState::DownLeft | LayoutState::DownRight => {
+                vec![
+                    vec![EdgeShape::id(&front.side_line),  EdgeShape::id(&back.side_line)  ],
+                    vec![EdgeShape::id(&front.corner),     EdgeShape::id(&back.corner)     ],
```
I was writing in my previous PR that this is incorrect. `front.corner` and `back.corner` do not "occupy the same space", as docs tell. Front corner is the part that is visible next to node, while back corner is a very short part which fills the hole between the shape and `front.corner`. They are visible at the same time, always. - see the definition of the shapes - they differ . One is `ring - node_shape` and one is `ring * node_shape`.

MichaelMauderer replied:
I've added a note about this. It doesn't really matter as there is no way a focus split can create a self-intersection on those shapes. We can just handle them as part of the same  shape group.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+impl ShapeViewEventsProxy {
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! { network
+            def on_mouse_over = source();
+            def on_mouse_out  = source();
+            def on_mouse_down = source();
+
```
don't use defs pls . Use <- instead

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+            def on_mouse_out  = source();
+            def on_mouse_down = source();
+
+            mouse_down <- on_mouse_down.constant(());
+            mouse_over <- on_mouse_over.constant(());
+            mouse_out  <- on_mouse_out.constant(());
```
the type of `on_mouse_down` is `()` so this constant doesn't make sense here. If you want to convert from `Source` to `Stream`, use `into` instead.

MichaelMauderer replied:
`on_mouse_down` here is an `frp::Source<display::object::Id>`.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+    pub source_attached : frp::Source<bool>,
+    pub redraw          : frp::Source,
+
+    pub hover_position  : frp::Source<Option<Vector2<f32>>>,
```
Few lines above we are using `Vector2` and here we are using `Vector2<f32>` - lets make it consistent and lets skip the parametrization

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -432,16 +1125,39 @@ impl Edge {
+            eval shape_events.on_mouse_over ((id) {
+                hover_target.set(Some(*id));
+            });
+            eval shape_events.on_mouse_out  ((_)  {
```
use eval_

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -432,16 +1125,39 @@ impl Edge {
+            eval input.source_height   ((t) source_height.set(*t));
+            eval input.hover_position  ((t) hover_position.set(*t));
+
+            eval shape_events.on_mouse_over ((id) {
+                hover_target.set(Some(*id));
+            });
```
make such things one-liners and do not use curly braces here

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -513,11 +1253,33 @@ impl EdgeModelData {
+        let source_node_half_height = self.source_height.get() / 2.0;
+        let source_node_circle      = Vector2(node_half_width- source_node_half_height, 0.0);
+        let source_node_radius      = source_node_half_height;
+
+        // === Update Highlights ===
+
```
such things should have 2 enters above and one below according to our style guide

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -573,6 +1335,16 @@ impl EdgeModelData {
         let is_down        = if is_flat_side {downward_flat} else {downward_far};
 
+        // === Layout State ===
+        // Initial guess at our layout. Will be refined for some edge cases when we have more
+        // layout informaiton.
```
What does it mean "when we have more layout information"? I mean, this should be explained better here - what is to be refined and what info is missing at this point? Just add few words about it - with an example info that is missing here that will be known later

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -573,6 +1335,16 @@ impl EdgeModelData {
+        // Initial guess at our layout. Will be refined for some edge cases when we have more
+        // layout informaiton.
+        let state = match (is_down,(side < 0.0)) {
+            (true,true) => LayoutState::DownLeft,
+            (true,false) => LayoutState::DownRight,
+            (false,true) => LayoutState::UpLeft,
+            (false,false) => LayoutState::UpRight,
+        };
```
alignment

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -884,7 +1665,206 @@ impl EdgeModelData {
+        delta_y > 0.0 && delta_y < MIN_SOURCE_TARGET_DIFFERENCE_FOR_Y_VALUE_DISCRIMINATION
+    }
+
+    // FIXE: several lines are too long.
```
please fix this "FIXE" (my keyboard doesnt like the key "m" :( )

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -884,7 +1665,206 @@ impl EdgeModelData {
+        }
+    }
+
+    /// These corrections are needed as sometimes shapes are in places that lead to inconsistent
+    /// results, e.g., the side line leaving the node from left/right or right/left. The shape
+    /// itself does not have enough information about its own placement to determine which end
+    /// is pointed towards the `Target` or `Source` part of the whole edge. So we need to account
+    /// for these here based on the specific layout state we are in.
```
Could you please describe an example case preferably with ascii-art here? It would be sooooo helpful to understand it. Now I understand what is happening here, but I bet in couple of months this will be hard to follow.

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -884,7 +1665,206 @@ impl EdgeModelData {
+
+        let flip = 2.0 * RIGHT_ANGLE;
+
+        // These rules are derived from the algorithm in `redraw`. In some layout configurations
+        // shapes are inverted top/down or left/right and we need to apply the appropriate
+        // corrections here. Sometimes these are just the side-effect of some layouting mechanics
+        // without visual justification (e.g., the `PortLine` sometimes ends up with a negative
+        // height and is thus flipped upside down.
```
super cool

wdanilo commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -884,7 +1665,206 @@ impl EdgeModelData {
+        let semantic_split = SemanticSplit::new(&self,snap_data.target_shape_id).ok_or(())?;
+        let angle          = self.cut_angle_for_shape(snap_data.target_shape_id,position,part).ok_or(())?;
+
+        // Completely disable/enable focus for shapes that are not close the split base don their
```
there are some mistakes in words here

MichaelMauderer commented on src/rust/ide/lib/view/src/graph_editor.rs:
```diff
@@ -1617,25 +1682,101 @@ fn new_graph_editor(app:&Application) -> GraphEditor {
+        Some(model.new_edge_from_output(&edge_mouse_down,&edge_over,&edge_out))
+    })).unwrap();
+    new_input_edge <- create_edge_from_input.map(f!([model,edge_mouse_down,edge_over,edge_out]((target)){
+        if model.is_node_connected_at_input(target.node_id,target.port.to_vec()) {
+            return None
+        };
+        Some(model.new_edge_from_input(&edge_mouse_down,&edge_over,&edge_out))
```
Doesn't seem to be implemented? `method not found in 'graph_editor::EdgeId'`

MichaelMauderer commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
 const MOUSE_OFFSET       : f32 = 2.0;
 const NODE_PADDING       : f32 = node::SHADOW_SIZE;
-const PADDING            : f32 = 4.0;
+const PADDING            : f32 = 4.0 + HOVER_EXTENSION;
```
If I remember correctly this was to void the hover area beeing clipped at the edges of the sprite.

MichaelMauderer commented on src/rust/ide/lib/view/src/graph_editor/component/edge.rs:
```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+const PADDING            : f32 = 4.0 + HOVER_EXTENSION;
 const RIGHT_ANGLE        : f32 = std::f32::consts::PI / 2.0;
 
+// TODO: Maybe find a better name.
+const MIN_SOURCE_TARGET_DIFFERENCE_FOR_Y_VALUE_DISCRIMINATION : f32 = 45.0;
```
This is related to the issue, that we want to split edges based on the y-value. But if they are next to each other that doesn't work. So we need to switch between "use y-value" and "use euclidian distance" when the y-distance is too small. This value is the threshold for that switch.

MichaelMauderer commented on build/run.js:
```diff
@@ -127,7 +127,7 @@ commands.build.rust = async function(argv) {
         console.log('Checking the resulting WASM size.')
         let stats = fss.statSync(paths.dist.wasm.mainOptGz)
-        let limit = 3.54
+        let limit = 3.58
```
After re-basing the increase is from  3.59 (main) to 3.62 for this PR. No libraries added.
